src/react/elements.js

```diff
@@ -0,0 +1,142 @@
+      invariant(props instanceof ObjectValue);
+      Properties.Set(realm, props, "children", value, true);
+    } else if (name === "key" && value !== realm.intrinsics.null) {
+      key = computeBinary(realm, "+", realm.intrinsics.emptyString, value);
+    } else if (name === "ref") {
+      ref = value;
+    } else if (name !== "__self" && name !== "__source") {
+      invariant(props instanceof ObjectValue);
+      Properties.Set(realm, props, name, value, true);
+    }
+  };
+
+  if (
+    (config instanceof AbstractObjectValue && config.isPartialObject()) ||
+    config instanceof AbstractValue ||
+    (config instanceof ObjectValue && config.isPartialObject() && config.isSimpleObject())
+  ) {
+    // if we have defaultProps, we need to create a new merge of the objects along with our config
+    if (defaultProps !== realm.intrinsics.undefined || children !== realm.intrinsics.undefined) {
+      if (objectHasNoPartialKeyAndRef(realm, config)) {
+        let args = [];
+        if (defaultProps !== realm.intrinsics.undefined) {
+          args.push(defaultProps);
+        }
+        args.push(config);
+        // create a new props object that will be the target of the Object.assign
+        props = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+        // as this is "props" that is abstract, we need to make it partial and simple
+        props.makePartial();
+        props.makeSimple();
+        // props objects also don't have a key and ref, so we remove them
+        deleteRefAndKeyFromProps(realm, props);
+        let types = new TypesDomain(FunctionValue);
+        let values = new ValuesDomain();
+
+        // get the global Object.assign
+        let globalObj = Get(realm, realm.$GlobalObject, "Object");
+        invariant(globalObj instanceof ObjectValue);
+        let objAssign = Get(realm, globalObj, "assign");
+        invariant(realm.generator);
+
+        realm.generator.derive(types, values, [objAssign, props, ...args], ([methodNode, ..._args]) => {
+          return t.callExpression(methodNode, ((_args: any): Array<any>));
+        });
+
+        if (children !== realm.intrinsics.undefined) {
+          setProp("children", children);
+        }
+      } else {
+        // if either are abstract, this will impact the reconcilation process
+        // and ultimately prevent us from folding ReactElements properly
+        let diagnostic = new CompilerDiagnostic(
+          `unable to evaluate "key" and "ref" on a ReactElement due to an abstract config passed to createElement`,
+          realm.currentLocation,
+          "PP0025",
+          "FatalError"
+        );
+        realm.handleError(diagnostic);
+        if (realm.handleError(diagnostic) === "Fail") throw new FatalError();
+      }
+    } else {
+      props = config;
+    }
+  } else {
+    if (config instanceof ObjectValue) {
+      for (let [propKey, binding] of config.properties) {
+        if (binding && binding.descriptor) {
```

Comment:
I think this needs `&& binding.descriptor.enumerable` to fix https://github.com/facebook/prepack/pull/1447#issuecomment-367359080.